source/LottieToWinComp/LottieToWinCompTranslator.cs

```
@@ -268,121 +334,229 @@ IEnumerable<Visual> VisualsAndShapesToVisuals(TranslationContext context, IEnume
+                        {
+                            maskMode = mask.Mode;
+                        }
+
+                        if (mask.Mode == maskMode)
+                        {
+                            var maskSpriteShape = CreateSpriteShape();
+                            maskSpriteShape.Geometry = compositionPathGeometry;
+
+                            // The mask geometry needs to be colored with something so that it can be used
+                            // as a mask.
+                            maskSpriteShape.FillBrush = CreateColorBrush(LottieData.Color.Black);
+
+                            maskContainerShape.Shapes.Add(maskSpriteShape);
+                        }
+                        else
+                        {
+                            _issues.MaskModesDoNotMatch();
+                        }
+                    }
                 }
+            }
+
+            return maskMode;
+        }
 
-                // Add a mask as a clip path if it can support the mask properties
-                if (!mask.Inverted &&
-                    mask.Opacity.InitialValue == 100 &&
-                    !mask.Opacity.IsAnimated &&
-                    mask.Mode == Mask.MaskMode.Additive &&
-                    layer.Masks.Length == 1)
+        // Translate a mask into shapes for a shape visual. The mask is applied to the visual to be masked
+        // using the VisualSurface. The VisualSurface can take the rendered contents of a visual tree and
+        // use it as a brush. The final masked result is achieved by taking the visual to be masked, putting
+        // it into a VisualSurface, then taking the mask and putting that in a VisualSurface and then combining
+        // the result with a composite effect.
+        Visual TranslateAndApplyMasks(
+            TranslationContext context,
+            Layer layer,
+            Visual visualToBeMasked)
+        {
+            if (layer.Masks != null &&
```

>layer.Masks != [](start = 16, length = 14)

Can we give Layer a default empty list of masks so we never have to check it for null?